modules/renter/download.go

```diff
@@ -0,0 +1,150 @@
+	return r.gateway.RPC(piece.HostIP, "RetrieveFile", func(conn modules.NetConn) (err error) {
+		// Send the id of the contract for the file piece we're requesting. The
+		// response will be the file piece contents.
+		if err = conn.WriteObject(piece.ContractID); err != nil {
+			return
+		}
+
+		// Create the file on disk.
+		file, err := os.Create(path)
+		if err != nil {
+			return
+		}
+		defer file.Close()
+
+		// Simultaneously download file and calculate its Merkle root.
+		tee := io.TeeReader(
+			// use a LimitedReader to ensure we don't read indefinitely
+			io.LimitReader(conn, int64(piece.Contract.FileSize)),
+			// each byte we read from tee will also be written to file
+			file,
+		)
+		merkleRoot, err := crypto.ReaderMerkleRoot(tee)
+		if err != nil {
+			return
+		}
+
+		if merkleRoot != piece.Contract.FileMerkleRoot {
+			return errors.New("host provided a file that's invalid")
+		}
+
+		return
+	})
+}
+
+// Download downloads a file. Mutex conventions are broken to prevent doing
+// network communication with io in place.
+func (r *Renter) Download(nickname, filename string) error {
+	// Grab the set of pieces we're downloading.
+	r.mu.RLock()
+	var pieces []FilePiece
+	_, exists := r.files[nickname]
+	if !exists {
+		r.mu.RUnlock()
+		return errors.New("no file of that nickname")
+	}
+	for _, piece := range r.files[nickname].pieces {
+		if piece.Active {
+			pieces = append(pieces, piece)
+		}
+	}
+	r.mu.RUnlock()
+
+	// Create an object for the download in the download queue.
+	r.mu.Lock()
+	downloadIndex := len(r.downloadQueue)
+	r.downloadQueue = append(r.downloadQueue, Download{
+		completed:   false,
+		destination: filename,
+		nickname:    nickname,
+
+		renter: r,
+	})
+	r.mu.Unlock()
+
+	// We only need one piece, so iterate through the hosts until a download
+	// succeeds.
+	go func() {
+		for i := 0; i < downloadAttempts; i++ {
+			for _, piece := range pieces {
+				downloadErr := r.downloadPiece(piece, filename)
+				if downloadErr == nil {
+					// Mark the download as complete.
+					r.mu.Lock()
+					r.downloadQueue[downloadIndex].completed = true
+					r.mu.Unlock()
+					return
+				}
+			}
+
+			// This iteration failed, no hosts returned the piece. Try again
+			// after waiting a random amount of time.
+			randSource := make([]byte, 1)
+			rand.Read(randSource)
+			time.Sleep(time.Second * time.Duration(i) * time.Duration(i) * time.Duration(randSource[0]))
+		}
+	}()
+
+	return nil
+}
+
+// DownloadQueue returns the list of downloads in the queue.
+func (r *Renter) DownloadQueue() []modules.DownloadInfo {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	downloads := make([]modules.DownloadInfo, len(r.downloadQueue))
+	for i := 0; i < len(r.downloadQueue); i++ {
+		downloads[i] = &r.downloadQueue[i]
+	}
+	return downloads
+}
+
+// Getter for the completed status of the download.
+func (d *Download) Completed() bool {
+	d.renter.mu.RLock()
+	defer d.renter.mu.RUnlock()
+	return d.completed
+}
```

Comment:
Alternatively I could give them their own lock, but that seemed excessive especially if using safelocks.